NativePromise should provide a safe way to pass some argument types

jyavenard opened the pull request:
#### 0345168e97b5e2c1fa074c8b5c9d721e23ae60ce
<pre>
NativePromise should provide a safe way to pass some argument types
<a href="https://bugs.webkit.org/show_bug.cgi?id=262437">https://bugs.webkit.org/show_bug.cgi?id=262437</a>
rdar://116274506

Reviewed by Youenn Fablet.

We expand the scope of CrossThreadCopier so that it can always deal with
r-values, including those without an isolatedCopy() method.
Add specialisation for void, std::tuple and AtomString.
We allow using non-threadsafe refcount when using only r-value as the object
can be safely moved around.

NativePromise aims to add a layer between the producer and the consumer such
that the producer doesn&apos;t have to know the threading model of the consumer.
To achieve this, we needed to have a way to guarantee that the object returned
by the producer is usable on any threads. By using this expanded CrossThreadCopier
in NativePromise we achieve that goal.
It is now safe to have NativePromise using Strings, AtomString and any
supported composite objects using those (tuple, variant, pair, expected,
Vector, optional, HashMap, HashSet)
Should more specialisation be needed, they should be added to CrossThreadCopier.

Non Exclusive NativePromise requires the programmer to explicitly define
the behaviour: that is always use CrossThreadCopy or never do.

Add API tests.

* Source/WTF/wtf/CrossThreadCopier.cpp:
* Source/WTF/wtf/CrossThreadCopier.h:
* Source/WTF/wtf/Forward.h:
* Source/WTF/wtf/NativePromise.h:
(WTF::operator|):
(WTF::operator&amp;):
* Source/WTF/wtf/TypeTraits.h:
* Source/WebCore/Modules/webaudio/AsyncAudioDecoder.cpp: Make use of the ability of
rejecting with an Exception object.
(WebCore::AsyncAudioDecoder::decodeAsync):
* Source/WebCore/Modules/webaudio/AsyncAudioDecoder.h: Update definition of DecodingTaskPromise
* Source/WebCore/Modules/webaudio/BaseAudioContext.cpp:
(WebCore::BaseAudioContext::decodeAudioData):
* Source/WebCore/dom/ExceptionOr.h:
* Source/WebKit/Scripts/webkit/messages.py:
(message_to_struct_declaration):
* Source/WebKit/Scripts/webkit/tests/TestWithCVPixelBufferMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithImageDataMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithLegacyReceiverMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithSemaphoreMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithStreamMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithSuperclassMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithoutAttributesMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithoutUsingIPCConnectionMessages.h:
* Tools/TestWebKitAPI/Tests/WTF/NativePromise.cpp:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/268880@main">https://commits.webkit.org/268880@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6aa0100da64ba8d23dda382c6744c216e246e0c2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20858 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/21266 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21926 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22748 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19435 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/21095 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24504 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21442 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/20722 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/21080 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20847 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/18111 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23602 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/21045 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/18011 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18924 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/25220 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/18168 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19091 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/19107 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/23130 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/20280 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19688 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16719 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/24268 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18929 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/5764 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5026 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/23258 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/25530 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19503 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5589 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

squelart commented on Source/WTF/wtf/CrossThreadCopier.cpp:
```diff
@@ -44,11 +44,6 @@ static_assert((std::is_same<RefPtr<CopierThreadSafeRefCountedTest>, CrossThreadC
 static_assert((std::is_same<RefPtr<CopierThreadSafeRefCountedTest>, CrossThreadCopier<CopierThreadSafeRefCountedTest*>::Type>::value), "RawPointerTest");
```
Typos in the commit description (not this line of code) :
- recount -> refcount
- supportde -> supported

squelart replied:
"supportde" was fixed, but not "recount" yet.

squelart commented on Source/WTF/wtf/CrossThreadCopier.h:
```diff
@@ -94,33 +104,74 @@ template<typename T> struct CrossThreadCopierBase<false, true, T> {
+
+// Can only be moved
+template<typename T> struct CrossThreadCopierBase<false, false, Ref<T>> {
+
```
Remove empty line.

squelart commented on Source/WTF/wtf/CrossThreadCopier.h:
```diff
@@ -94,33 +104,74 @@ template<typename T> struct CrossThreadCopierBase<false, true, T> {
+
+// Can only be moved
+template<typename T> struct CrossThreadCopierBase<false, false, RefPtr<T>> {
+
```
Remove empty line.

squelart commented on Source/WTF/wtf/CrossThreadCopier.h:
```diff
@@ -94,33 +104,74 @@ template<typename T> struct CrossThreadCopierBase<false, true, T> {
+    {
+        return String(source).isolatedCopy();
+    }
+    static Type copy(Type&& source)
```
Since this is the specialisation for `AtomString`, presumably the `copy` functions will be called with an `AtomString`, so by taking a `String` reference you'll force a conversion *before* you enter these functions, is that intended? Shouldn't you take an `AtomString` reference and then in the functions do the conversion to a String and isolate-copy it?

jyavenard replied:
you're entirely right.

squelart commented on Source/WTF/wtf/CrossThreadCopier.h:
```diff
@@ -241,11 +294,58 @@ template<typename... Types> struct CrossThreadCopierBase<false, false, std::vari
+    static Type copy(const Type& source)
+    {
+        if (source.has_value()) {
+            if constexpr (std::is_void<T>::value)
```
Nit: `std::is_void_v<T>` (and no `::value`).

squelart commented on Source/WTF/wtf/CrossThreadCopier.h:
```diff
@@ -241,11 +294,58 @@ template<typename... Types> struct CrossThreadCopierBase<false, false, std::vari
+    static Type copy(Type&& source)
+    {
+        if (source.has_value()) {
+            if constexpr (std::is_void<T>::value)
```
Nit: `std::is_void_v<T>` (and no `::value`).

squelart reviewed (approved):
r+ with nits.
I'm not an official reviewer yet, so please get a second opinion.

youennf commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -340,7 +342,7 @@ class NativePromise final : public NativePromiseBase, public ConvertibleToNative
         PROMISE_LOG(resolveSite, " resolving ", *this);
         ASSERT(isNothing());
-        m_result = std::forward<ResolveValueType_>(resolveValue);
+        m_result = crossThreadCopy(std::forward<ResolveValueType_>(resolveValue));
```
This is good to cross thread copy since NativePromise may be destroyed in any thread.
But I am not sure this is enough, we probably want to ensure that m_result is cross thread copied for every then callback, since each one of them may use m_result on a different queue.
In that sense, I would remove NativePromise::result().

If we add that, it would be nice to keep the optimisation for the one then callback native promise.

jyavenard replied:
No, it doesn't have to be.

If IsExclusive==true, it is moved.

If IsExclusive==false : the object is passed via const reference; and the object is guaranteed to exist and not be modified while the callback is being called. Even if accessed from a different queue, an originally isolated object is usable (at least for all the objects we currently have in the tree).

As such, reading (or copying) the result is safe, regardless of the value of IsExclusive.

That only thing that isn't is taking a refcount of a non thread-safe refcounted value; but for that there's not much we can do as and is already protected through runtime assertions

jyavenard replied:
> But I am not sure this is enough, we probably want to ensure that m_result is cross thread copied for every then callback, since each one of them may use m_result on a different queue. In that sense, I would remove NativePromise::result().

Ok.

If I fully understand your aim, you're hoping that for object that have non-thread safe  refcounting you would be able to use them safely across multiple-threads through some hidden, under the hood mechanism.

And that is not achievable because at the time you resolve or reject a promise, it may not have any ThenCommand attached yet. A NativePromise can be then()ed on any threads. 
At the time you then()ed you may not be on the producer's thread anymore, you've lost the ability to safely perform another ref() 

Consider the following:
```
Ref<NonExclusiveNativePromise> p = promiseProducer(); /* returned promise has already been resolved */
m_workQueue->dispatch[p = WTFMove(p)] {
  p->whenSettled(RunLoop::main(), [](const NonExclusiveNativePromise::Result& result) { /* do something else */ }
}
```

As it is, the reference that the callback receives to the resolve/reject object is now immutable and read-only. You can perform a copy, but you can't take an extra ref() to it in a safe fashion. If you need to achieve this, then the object needs to have thread-safe refcounting. There's no other way about it.

We can store a `crossThreadCopy` at the time resolve/reject and pass that around, which is what this PR is doing.

This can be explained and documented, with the caveats listed, but C++ isn't Rust unfortunately, there are restrictions.

jyavenard replied:
As a follow-up patch, I can make the use of IsExclusive=false forbidden if the Resolve/RejectValueT contains a String or AtomString with a static_assert.

This will cover your concerns about misuse of the promise.

jyavenard replied:
I have implemented a new strategy with this latest PR.

You now define the behaviour with a PromiseOption enum.
There are two relevants:
WithCrossThreadCopy and WithoutCrossThreadCopy, if neither are set (the default) then it will automatically use crossThreadCopy on the resolve/reject value if one of the type used contains an AtomString or a method `isolatedCopy` 
Creating a non-exclusive NativePromise without setting either WithCrossThreadCopy or WithoutCrossThreadCopy is now forbidden.
if `WithCrossThreadCopy` is set, it will be up to the consumer to call crossThreadCopy() on the resolve/reject value provided as needed.

youennf commented on Tools/TestWebKitAPI/Tests/WTF/NativePromise.cpp:
```diff
@@ -794,10 +794,156 @@ TEST(NativePromise, MoveOnlyType)
+    AutoWorkQueue awq;
+    auto queue = awq.queue();
+
+    MyPromise::createAndResolve(String("hello"_s))->then(queue,
```
Let's use multiple queues in that test for the different then/whenSettled.
And let's keep some refs to val that we keep in the same queue via a dispatch so that the other callbacks execute while past callbacks hold a ref to string.

jyavenard replied:
ok. But all new then will have an object with a single refcount. that's the whole point

youennf reviewed (commented):
I would beef up the tests a bit and verify whether we need to do more cross thread copies at then callback call time.

squelart reviewed (approved):
Nice! Still (unofficial) r+ from me.

youennf commented on Tools/TestWebKitAPI/Tests/WTF/NativePromise.cpp:
```diff
@@ -794,14 +794,230 @@ TEST(NativePromise, MoveOnlyType)
+    AutoWorkQueue awq;
+    auto queue = awq.queue();
+
+    MyPromise::createAndResolve("hello"_s)->then(queue,
```
It would be good to add a flow where we create a promise, call then and resolve the promise since the code path is slightly different. Both for resolve and reject.

jyavenard replied:
There's a few of those in the current API test. 
I'll add one.

youennf commented on Tools/TestWebKitAPI/Tests/WTF/NativePromise.cpp:
```diff
@@ -794,14 +794,230 @@ TEST(NativePromise, MoveOnlyType)
+
+    {
+        // If you replace PromiseOption::WithCrossThreadCopy with PromiseOption::WithoutCrossThreadCopy, this test will crash due to the AtomString being deleted on the target queue.
+        using MyPromise = NativePromise<Expected<String, AtomString>, bool, PromiseOption::NonExclusive | PromiseOption::WithCrossThreadCopy>;
```
Given it is non exclusive, we should probably exercise its non exclusivity.
That said, I would still prefer we forbid this case for now until we have a proof this is both useful and not error prone.

youennf commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -207,13 +214,42 @@ enum class PromiseDispatchMode : uint8_t {
+    return a & static_cast<unsigned>(b);
+}
+
+// TODO: Determine a compile time the value of WithCrossThreadCopy, by identifying if ResolveValueT/RejectValueT contains
```
s/a/at/

jyavenard replied:
should remove, as I've done so now.

youennf commented on Source/WebCore/Modules/webaudio/AsyncAudioDecoder.cpp:
```diff
@@ -47,7 +47,7 @@ Ref<DecodingTaskPromise> AsyncAudioDecoder::decodeAsync(Ref<ArrayBuffer>&& audio
         callOnMainThread([audioData = WTFMove(audioData)] { });
         if (!audioBuffer)
-            return DecodingTaskPromise::createAndReject(EncodingError);
+            return DecodingTaskPromise::createAndReject(Exception { EncodingError, "Decoding failed"_s });
```
This is great !

youennf reviewed (approved):
LGTM.
I would still disable for now IsExclusive=false with a type whose CrossThreadCopier::IsNeeded is true.
AFAIK, we do not have a use case for it right now and it is a dangerous pattern in its current form.

jyavenard committed:
NativePromise should provide a safe way to pass some argument types
https://bugs.webkit.org/show_bug.cgi?id=262437
rdar://116274506

Reviewed by Youenn Fablet.

We expand the scope of CrossThreadCopier so that it can always deal with
r-values, including those without an isolatedCopy() method.
Add specialisation for void, std::tuple and AtomString.
We allow using non-threadsafe refcount when using only r-value as the object
can be safely moved around.

NativePromise aims to add a layer between the producer and the consumer such
that the producer doesn't have to know the threading model of the consumer.
To achieve this, we needed to have a way to guarantee that the object returned
by the producer is usable on any threads. By using this expanded CrossThreadCopier
in NativePromise we achieve that goal.
It is now safe to have NativePromise using Strings, AtomString and any
supported composite objects using those (tuple, variant, pair, expected,
Vector, optional, HashMap, HashSet)
Should more specialisation be needed, they should be added to CrossThreadCopier.

Non Exclusive NativePromise requires the programmer to explicitly define
the behaviour: that is always use CrossThreadCopy or never do.

Add API tests.

* Source/WTF/wtf/CrossThreadCopier.cpp:
* Source/WTF/wtf/CrossThreadCopier.h:
* Source/WTF/wtf/Forward.h:
* Source/WTF/wtf/NativePromise.h:
(WTF::operator|):
(WTF::operator&):
* Source/WTF/wtf/TypeTraits.h:
* Source/WebCore/Modules/webaudio/AsyncAudioDecoder.cpp: Make use of the ability of
rejecting with an Exception object.
(WebCore::AsyncAudioDecoder::decodeAsync):
* Source/WebCore/Modules/webaudio/AsyncAudioDecoder.h: Update definition of DecodingTaskPromise
* Source/WebCore/Modules/webaudio/BaseAudioContext.cpp:
(WebCore::BaseAudioContext::decodeAudioData):
* Source/WebCore/dom/ExceptionOr.h:
* Source/WebKit/Scripts/webkit/messages.py:
(message_to_struct_declaration):
* Source/WebKit/Scripts/webkit/tests/TestWithCVPixelBufferMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithImageDataMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithLegacyReceiverMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithSemaphoreMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithStreamMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithSuperclassMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithoutAttributesMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithoutUsingIPCConnectionMessages.h:
* Tools/TestWebKitAPI/Tests/WTF/NativePromise.cpp:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/268880@main
